tests/integration/construct_component_provider_propagation/go/main.go

```diff
@@ -0,0 +1,67 @@
+//go:build !all
+// +build !all
+
+package main
+
+import (
+	"reflect"
+
+	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
+)
+
+func main() {
+	pulumi.Run(func(ctx *pulumi.Context) error {
+		provider, err := NewRandomProvider(ctx, "explicit")
+		if err != nil {
+			return err
+		}
+
+		if _, err := NewComponent(ctx, "uses_default", nil); err != nil {
+			return err
+		}
+
+		if _, err := NewComponent(ctx, "uses_provider", nil, pulumi.Provider(provider)); err != nil {
+			return err
+		}
+
+		if _, err := NewComponent(ctx, "uses_providers", nil, pulumi.Providers(provider)); err != nil {
+			return err
+		}
+
+		return nil
+	})
+}
+
+type RandomProvider struct {
+	pulumi.ProviderResourceState
+}
+
+func NewRandomProvider(ctx *pulumi.Context, name string) (*RandomProvider, error) {
+	var provider RandomProvider
+	err := ctx.RegisterResource("pulumi:providers:testprovider", "explicit", nil, &provider)
+	return &provider, err
+}
+
+type FooComponent struct {
+	pulumi.ResourceState
+
+	Result pulumi.StringOutput `pulumi:"result"`
+}
+
+func NewComponent(ctx *pulumi.Context, name string, args *FooComponentArgs, opts ...pulumi.ResourceOption) (*FooComponent, error) {
```

Comment:
Nit: `NewFooComponent`

FWIW, I would have called it `Component` to match the type token rather than `FooComponent`, but my care level for the naming in these tests is pretty low, so feel free to leave as-is.

Reply:
Renamed to Component